pennylane/interfaces/execution.py

```diff
@@ -119,6 +122,29 @@ def wrapper(tapes, **kwargs):
             if hashes[i] in cache:
                 # Tape exists within the cache, store the cached result
                 cached_results[i] = cache[hashes[i]]
+
+                # Introspect the set_shots decorator of the input function:
+                #   warn the user in case of finite shots with cached results
+                finite_shots = False
+
+                closure = inspect.getclosurevars(fn).nonlocals
+                if "original_fn" in closure:  # deal with expand_fn wrapper above
+                    closure = inspect.getclosurevars(closure["original_fn"]).nonlocals
+
+                # retrieve the captured context manager instance (for set_shots)
+                if "self" in closure and isinstance(closure["self"], _GeneratorContextManager):
+                    # retrieve the shots from the arguments or device instance
+                    if closure["self"].func.__name__ == "set_shots":
+                        dev, shots = closure["self"].args
+                        shots = dev.shots if shots is False else shots
+                        finite_shots = isinstance(shots, int)
+
+                if finite_shots:
+                    warnings.warn(
+                        "Cached execution with finite shots detected: note that the shot noise "

```

Comment:
Could we not have different samples produced with the same shot noise and so it's the samples that are the same? 

Also, it could be worth mentioning two things in the warning:
* This situation arises because the quantum function had the same operations,  measurements and trainable parameters (as per producing the tape hash);
* Doing `cache=False` should turn off this behaviour (likely this is unintended).